src/pudl/output/ferc1.py

```diff
@@ -2774,3 +2775,239 @@ def nodes_to_df(calc_forest: nx.DiGraph, nodes: list[NodeId]) -> pd.DataFrame:
     except AttributeError:
         tags = pd.DataFrame()
     return pd.concat([index, tags], axis="columns")
+
+
+def _propagate_tags_leafward(
+    annotated_forest: nx.DiGraph, leafward_inherited_tags: list[str]
+) -> nx.DiGraph:
+    existing_tags = nx.get_node_attributes(annotated_forest, "tags")
+    for node, parent_tags in existing_tags.items():
+        descendants = nx.descendants(annotated_forest, node)
+        descendant_tags = {
+            desc: {
+                "tags": {
+                    tag_name: parent_tags[tag_name]
+                    for tag_name in leafward_inherited_tags
+                    if tag_name in parent_tags
+                }
+                | existing_tags.get(desc, {})
+            }
+            for desc in descendants
+        }
+        nx.set_node_attributes(annotated_forest, descendant_tags)
+    return annotated_forest
+
+
+def _propagate_tag_rootward(
+    annotated_forest: nx.DiGraph, node, tag_name: Literal["in_rate_base"]
+) -> str:
+    """Set the tag for nodes when all of its children have same tag.
+
+    This function returns the value of a tag, but also sets node attributes
+    down the tree when all children of a node share the same tag.
+    """
+
+    def _get_tag(annotated_forest, node, tag_name):
+        return annotated_forest.nodes.get(node, {}).get("tags", {}).get(tag_name, pd.NA)
+
+    logger.info(f"propagaging tags leafward from {node}")
```

Comment:
Here's a non-recursive solution that uses `nx.topological_generations()`. I sort of think it's a little easier to read, since the flow is:

```
for each generation (starting from last):
    for each node in generation:
        set node tags, if children agree
```

We could skip the *very* last generation here since we know they have no children.

```diff
diff --git a/src/pudl/output/ferc1.py b/src/pudl/output/ferc1.py
index ecd6bfae3..4c9773bc8 100644
--- a/src/pudl/output/ferc1.py
+++ b/src/pudl/output/ferc1.py
@@ -1,5 +1,6 @@
 """A collection of denormalized FERC assets and helper functions."""
 import importlib
 import re
 from copy import deepcopy
 from functools import cached_property
@@ -2243,7 +2244,7 @@ class XbrlCalculationForestFerc1(BaseModel):
         annotated_forest = _propagate_tags_leafward(annotated_forest, ["in_rate_base"])
         # Rootward propagation
         root_node = self.roots(annotated_forest)[0]
-        _ = _propagate_tag_rootward(annotated_forest, root_node, "in_rate_base")
+        _propagate_tag_rootward(annotated_forest, "in_rate_base")
         # Correction Records
         annotated_forest = _propagate_tags_to_corrections(annotated_forest)
         return annotated_forest
@@ -2799,8 +2800,8 @@ def _propagate_tags_leafward(
 
 
 def _propagate_tag_rootward(
-    annotated_forest: nx.DiGraph, node, tag_name: Literal["in_rate_base"]
-) -> str:
+    annotated_forest: nx.DiGraph, tag_name: Literal["in_rate_base"]
+) -> None:
     """Set the tag for nodes when all of its children have same tag.
 
     This function returns the value of a tag, but also sets node attributes
@@ -2808,37 +2809,34 @@ def _propagate_tag_rootward(
     """
 
     def _get_tag(annotated_forest, node, tag_name):
-        return annotated_forest.nodes.get(node, {}).get("tags", {}).get(tag_name, pd.NA)
-
-    logger.info(f"propagaging tags leafward from {node}")
-    tag = pd.NA
-    # i'm a leaf so i stop looking or
-    # if i have a value you don't need to keep looking at this node's childern
-    if not list(annotated_forest.successors(node)) or not pd.isna(
-        _get_tag(annotated_forest, node, tag_name)
-    ):
-        tag = _get_tag(annotated_forest, node, tag_name)
-
-    else:
-        child_tags = set()
-        for child_node in annotated_forest.successors(node):
-            if not child_node.xbrl_factoid.endswith("_correction"):
-                child_tags.add(
-                    _propagate_tag_rootward(annotated_forest, child_node, tag_name)
-                )
-        # if all the children tags are the same and non-null
-        if (len(child_tags) == 1) and {t for t in child_tags if not pd.isna(t)}:
-            new_node_tag = child_tags.pop()
-            # actually assign the tag here but don't wipe out any other tags
-            existing_tags = nx.get_node_attributes(annotated_forest, "tags")
-            node_tags = {
-                node: {"tags": {tag_name: new_node_tag} | existing_tags.get(node, {})}
+        return annotated_forest.nodes.get(node, {}).get("tags", {}).get(tag_name)
+
+    generations = list(nx.topological_generations(annotated_forest))
+    for gen in reversed(generations):
+        untagged_nodes = {
+            node_id
+            for node_id in gen
+            if _get_tag(annotated_forest, node_id, tag_name) is None
+        }
+        for parent_node in untagged_nodes:
+            child_tags = {
+                _get_tag(annotated_forest, c, tag_name)
+                for c in annotated_forest.successors(parent_node)
+                if not c.xbrl_factoid.endswith("_correction")
             }
-            nx.set_node_attributes(annotated_forest, node_tags)
-            tag = new_node_tag
-        # elif the children disagree then the node's tag shouldn't be set and
-        # the og null tag should be returned
-    return tag
+            non_null_tags = child_tags - {None}
+            # sometimes, all children can share same tag but it's null.
+            if len(child_tags) == 1 and non_null_tags:
+                # actually assign the tag here but don't wipe out any other tags
+                new_node_tag = non_null_tags.pop()
+                existing_tags = nx.get_node_attributes(annotated_forest, "tags")
+                node_tags = {
+                    parent_node: {
+                        "tags": {tag_name: new_node_tag}
+                        | existing_tags.get(parent_node, {})
+                    }
+                }
+                nx.set_node_attributes(annotated_forest, node_tags)
 
 
 def _propagate_tags_to_corrections(annotated_forest: nx.DiGraph) -> nx.DiGraph:
```


It would probably be good, also, to break some of this logic out into its own little named functions - e.g. "do the children agree, on a non-null value?" is a surprisingly gnarly thing to figure out, and it seems very likely that we'd change the implementation in the future. By pulling it out we'd save a little time puzzling over "what code decides if the children agree?" in the future.